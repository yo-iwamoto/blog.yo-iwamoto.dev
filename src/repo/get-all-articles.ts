import { getServerSideEnv } from "@/config/env";
import { ArticleListItem } from "@/model/article-list-item";
import { cache } from "react";
import { cdnClient } from "./client";
import type { Content } from "./client";

const articleModelUid = "article";

async function getAllArticlesFn(): Promise<ArticleListItem[]> {
  const res = await cdnClient.getContents<
    {
      title: string;
      slug: string;
      tags: { name: string; slug: string }[];
    } & Content
  >({
    appUid: getServerSideEnv().newtAppUid,
    modelUid: articleModelUid,
    query: {
      order: ["-publishDate"],
      select: ["title", "slug", "tags", "_sys"],
    },
  });

  const articles = res.items.map(
    (item) =>
      new ArticleListItem(
        item.title,
        item.slug,
        item.tags,
        new Date(item._sys.raw.firstPublishedAt),
      ),
  );

  return articles;
}

export const getAllArticles = cache(getAllArticlesFn);
