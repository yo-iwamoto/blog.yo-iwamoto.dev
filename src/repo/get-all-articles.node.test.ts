import { getAllArticles } from "./get-all-articles";

const { getContents } = vi.hoisted(() => ({
  getContents: vi.fn(async () => ({
    items: [
      {
        title: "Article 1",
        slug: "article-1",
        tags: [],
        _sys: {
          raw: {
            firstPublishedAt: "2021-01-01T00:00:00.000Z",
          },
        },
      },
    ],
  })),
}));

vi.mock("./client", () => ({ cdnClient: { getContents } }));

describe("getAllArticles", () => {
  it("取得した記事一覧が整形して返却されること", async () => {
    const res = await getAllArticles();

    expect(res).toHaveLength(1);
    expect(res[0].title).toBe("Article 1");
    expect(res[0].slug).toBe("article-1");
    expect(res[0].tags).toHaveLength(0);
    expect(res[0].publishedAt).toStrictEqual(
      new Date("2021-01-01T00:00:00.000Z"),
    );
  });

  it("データ取得のクエリが変わらないこと", async () => {
    await getAllArticles();

    expect(getContents.mock.lastCall).toMatchSnapshot();
  });
});
