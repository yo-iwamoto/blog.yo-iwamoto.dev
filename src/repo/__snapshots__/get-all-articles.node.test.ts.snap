// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAllArticles > データ取得のクエリが変わらないこと 1`] = `
[
  {
    "appUid": "xxx",
    "modelUid": "article",
    "query": {
      "order": [
        "-publishDate",
      ],
      "select": [
        "title",
        "slug",
        "tags",
        "_sys",
      ],
    },
  },
]
`;
